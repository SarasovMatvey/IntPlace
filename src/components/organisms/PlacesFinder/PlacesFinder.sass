.places-finder
  display: grid
  grid-template-columns: repeat(3, 1fr)
  align-items: center
  row-gap: 2rem
  &__search-bar
    grid-column-start: 2
    justify-self: center
    align-self: center
  &__settings-btn
    justify-self: end
  &__request-settings
    grid-column-end: span 3
    .content
      display: none
    .content.active
      display: block
  &__list
    width: 100%
    grid-column-end: span 3
  &__card
    cursor: pointer
  &__request-settings-types
    display: grid
    grid-auto-flow: column
    column-gap: 1rem