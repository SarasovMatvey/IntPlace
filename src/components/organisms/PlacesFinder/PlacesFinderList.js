import { useEffect, useRef, useState } from "react";
import Masonry from "react-responsive-masonry";
import { useHistory } from "react-router-dom";
import sygicAxios from "../../../api/sygic";
import useOffset from "../../../utils/hooks/useOffset";
import PlaceCard from "../../atoms/PlaceCard";
import { DEFALUT_CATEGORIES_VALUES } from "./PlacesFinder";

function PlacesFinderList({
  searchQuery,
  countryId,
  categories,
  placesInPage,
  loadNewOffset,
}) {
  const history = useHistory();
  const placesFinderRef = useRef(null);
  const [isContentLoading, setIsContentLoading] = useState(false);
  const [shownPagesCount, setShownPagesCount] = useState(1);
  const [places, setPlaces] = useState([]);

  useOffset(placesFinderRef, loadNewOffset, () => {
    if (!isContentLoading) {
      setShownPagesCount(pagesCount => pagesCount + 1);
    }
  });

  useEffect(() => {
    setShownPagesCount(1);

    const url = getUrl({
      searchQuery,
      countryId,
      categories,
      limit: placesInPage,
    });

    sygicAxios(url).then(({ data: { data } }) => setPlaces(data.places));
  }, [searchQuery, countryId, categories, placesInPage]);

  useEffect(() => {
    const url = getUrl({
      searchQuery,
      countryId,
      categories,
      offset: (shownPagesCount - 1) * placesInPage,
      limit: placesInPage,
    });

    setIsContentLoading(true);
    sygicAxios(url.replace(/\s/g, "")).then(({ data: { data } }) => {
      setPlaces(places => [...places, ...data.places]);
      setIsContentLoading(false);
    });
  }, [shownPagesCount, searchQuery, countryId, categories, placesInPage]);

  return (
    <div className="places-finder__list" ref={placesFinderRef}>
      <Masonry gutter="2rem" columnsCount={4}>
        {places.map(placeInfo => (
          <PlaceCard
            className="places-finder__card"
            key={placeInfo.id}
            fluid
            title={placeInfo.name}
            poster={placeInfo.thumbnail_url}
            description={placeInfo.perex}
            placeId={placeInfo.id}
            onClick={e => goToPlaceInfoPage(e.currentTarget.dataset.id)}
          />
        ))}
      </Masonry>
    </div>
  );

  function goToPlaceInfoPage(placeId) {
    history.push(`place/${placeId}`);
  }

  function getUrl({
    searchQuery,
    countryId,
    categories,
    levels,
    offset,
    limit,
  }) {
    const queryParamStr = searchQuery
      ? `query=${searchQuery.replace(/s/g, "%20")}`
      : "";
    const parentsParamStr = `parents=country:${countryId}`;
    const categoriesParamStr = categories
      ? `categories=${categories.join("%7C")}`
      : `categories=${DEFALUT_CATEGORIES_VALUES.join("%7C")}`;
    const levelsParamStr = levels ? `levels=${levels}` : `levels=poi`;
    const offsetParamStr = offset ? `offset=${offset}` : "";
    const limitParamStr = limit ? `limit=${limit}` : "";

    const url = `/places/list?
      ${queryParamStr}&
      ${parentsParamStr}&
      ${categoriesParamStr}&
      ${levelsParamStr}&
      ${offsetParamStr}&
      ${limitParamStr}
    `;

    return url;
  }
}

export default PlacesFinderList;
