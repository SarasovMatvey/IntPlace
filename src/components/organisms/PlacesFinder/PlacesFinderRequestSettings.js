import { Accordion, Dropdown, Segment } from "semantic-ui-react";

function PlacesFinderRequestSettings({
  isOpen,
  countries,
  categories,
  defalutCountryValue,
  defalutCategoriesValues,
  onCountryChange,
  onCategoriesChange,
}) {
  return (
    <div className="places-finder__request-settings">
      <Accordion>
        <Accordion.Content active={isOpen}>
          <Segment className="places-finder__request-settings-types" padded>
            <Dropdown
              placeholder="Country"
              search
              selection
              options={countries}
              defaultValue={defalutCountryValue}
              onChange={onCountryChange}
            />
            <Dropdown
              placeholder="Categories"
              fluid
              multiple
              selection
              options={categories}
              defaultValue={defalutCategoriesValues}
              onChange={onCategoriesChange}
            />
          </Segment>
        </Accordion.Content>
      </Accordion>
    </div>
  );
}

export default PlacesFinderRequestSettings;
