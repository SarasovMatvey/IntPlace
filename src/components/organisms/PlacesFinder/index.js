export { default } from "./PlacesFinder.js";
