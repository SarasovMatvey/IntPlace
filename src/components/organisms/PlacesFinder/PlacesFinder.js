import { useRef, useState } from "react";
import { Button, Input } from "semantic-ui-react";
import PlacesFinderList from "./PlacesFinderList";
import PlacesFinderRequestSettings from "./PlacesFinderRequestSettings";
import "./PlacesFinder.sass";
import _ from "lodash";

const PLACES_IN_PAGE = 20;
const LOADING_NEW_PAGE_OFFSET = 100;
const CATEGORIES = ["eating", "traveling", "shopping"];
const COUNTRIES = {
  Russia: 37,
  USA: 43,
  Japan: 75,
};
const DEFALUT_COUNTRY_VALUE = 43;
export const DEFALUT_CATEGORIES_VALUES = ["eating", "traveling", "shopping"];

function PlacesFinder() {
  const placesFinderRef = useRef(null);
  const [isSettingsOpen, setIsSettingsOpen] = useState(false);

  const [preData, setPreData] = useState({
    searchQuery: null,
    countryId: DEFALUT_COUNTRY_VALUE,
    categories: DEFALUT_CATEGORIES_VALUES,
  });
  const [realData, setRealData] = useState({
    searchQuery: null,
    countryId: DEFALUT_COUNTRY_VALUE,
    categories: DEFALUT_CATEGORIES_VALUES,
  });

  return (
    <div className="places-finder" ref={placesFinderRef}>
      <Input
        className="places-finder__search-bar"
        size="huge"
        icon={{
          name: "search",
          circular: true,
          link: true,
          onClick: () => setRealData(_.cloneDeep(preData)),
        }}
        placeholder="Find Place..."
        onChange={e => {
          const value = e.currentTarget.value;
          setPreData(preData => ({
            ...preData,
            searchQuery: value,
          }));
        }}
      />
      <Button
        className="places-finder__settings-btn"
        circular
        icon="settings"
        onClick={() => setIsSettingsOpen(isSettingsOpen => !isSettingsOpen)}
      />
      <PlacesFinderRequestSettings
        isOpen={isSettingsOpen}
        countries={_.map(COUNTRIES, (countryId, countryName) => ({
          text: countryName,
          value: countryId,
        }))}
        defalutCountryValue={DEFALUT_COUNTRY_VALUE}
        onCountryChange={(_, { value }) =>
          setPreData(preData => ({ ...preData, countryId: value }))
        }
        categories={CATEGORIES.map(category => ({
          text: category,
          value: category,
        }))}
        defalutCategoriesValues={DEFALUT_CATEGORIES_VALUES}
        onCategoriesChange={(_, { value }) =>
          setPreData(preData => ({
            ...preData,
            categories: value,
          }))
        }
      />
      <PlacesFinderList
        countryId={realData.countryId}
        loadNewOffset={LOADING_NEW_PAGE_OFFSET}
        placesInPage={PLACES_IN_PAGE}
        searchQuery={realData.searchQuery}
        categories={realData.categories}
      />
    </div>
  );
}

export default PlacesFinder;
