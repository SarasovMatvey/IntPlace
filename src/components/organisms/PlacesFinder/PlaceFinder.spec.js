import PlacesFinder from "./PlacesFinder";

const setUp = props => shallow(<PlacesFinder {...props} />);

describe("PlacesFinder", () => {
  let component;

  beforeEach(() => {
    component = setUp();
  });

  it("should renders properly", () => {
    expect(component).toMatchSnapshot();
  });
});
