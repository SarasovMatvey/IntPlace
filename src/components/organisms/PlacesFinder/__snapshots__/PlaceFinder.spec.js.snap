// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesFinder should renders properly 1`] = `
<div
  className="places-finder"
>
  <Input
    className="places-finder__search-bar"
    icon={
      Object {
        "circular": true,
        "link": true,
        "name": "search",
        "onClick": [Function],
      }
    }
    onChange={[Function]}
    placeholder="Find Place..."
    size="huge"
    type="text"
  />
  <Button
    as="button"
    circular={true}
    className="places-finder__settings-btn"
    icon="settings"
    onClick={[Function]}
  />
  <PlacesFinderRequestSettings
    categories={
      Array [
        Object {
          "text": "eating",
          "value": "eating",
        },
        Object {
          "text": "traveling",
          "value": "traveling",
        },
        Object {
          "text": "shopping",
          "value": "shopping",
        },
      ]
    }
    countries={
      Array [
        Object {
          "text": "Russia",
          "value": 37,
        },
        Object {
          "text": "USA",
          "value": 43,
        },
        Object {
          "text": "Japan",
          "value": 75,
        },
      ]
    }
    defalutCategoriesValues={
      Array [
        "eating",
        "traveling",
        "shopping",
      ]
    }
    defalutCountryValue={43}
    isOpen={false}
    onCategoriesChange={[Function]}
    onCountryChange={[Function]}
  />
  <PlacesFinderList
    categories={
      Array [
        "eating",
        "traveling",
        "shopping",
      ]
    }
    countryId={43}
    loadNewOffset={100}
    placesInPage={20}
    searchQuery={null}
  />
</div>
`;
